### DISCUSSION TITLE
Duel Mimir-distributed not receiving the same metrices in Grafana

### DISCUSSION THREAD

**Discussion by akalita-conga** [grafana/mimir#5484] [Help and support]
We deployed two Mimir stack in the cluster (Mimir1 and Mimir2) using helm chart. 
We are using OTEL collector for sending cluster metrics into both Mimir equally without using any filter.
Both Mimirs are backed by two AWS S3.

The following components are deployed as part of the deployment,
  
 1. Mimir1
   compactor
   distributor
   ingester-zone-a-0                                                    
   nginx-546c8d                       
   querier
   query-frontend-7c464             
   rollout-operator-5b
   store-gateway-zone-a-0                           
 2. Mimir2
   chunks-cache
   compactor
   distributor
   index-cache
   ingester-zone
  metadata-cache
  nginx-587dd
  overrides-exporter
  querier-79767d
  query-frontend
  query-scheduler
  rollout-operator
  store-gateway

1. Mimir helm chart version
   a. Mimir1-
     appVersion: 2.9.0,version: 4.5.0

   b. Mimir2-
     appVersion: 2.7.1,version: 4.3.0

4. Perform Operations
   In Grafana two datasources configured mimir1/mimir2
5.Expected behavior
 Both mimir datasource should have same set of metrices in the grafana .
 
6.Environment
   Dev

- Infrastructure: [e.g., Kubernetes, bare-metal, laptop]
  EKS
- Deployment tool: [e.g., helm, jsonnet]
   Helm

7. Additional Context
 Both Mimir datasource is showing different set of metrices .To resolve the issue we have cleared the Mimir1 S3,cleared PVC  and redeployed again but the metrics data are still not same in the two datasource .

**Comment by dimitarvdimitrov** [comment]
Hi, can you show some examples of the metrics not being the same? Also in the list of deployed components in the two clusters seems very different, can you double check it, please?

**Comment by akalita-conga** [comment]
@dimitarvdimitrov 
Mimir1-No kube state metrics available
<img width="229" alt="image" src="https://github.com/grafana/mimir/assets/111333290/db94cfc1-5696-4507-9156-cbd1cd9f837f">
Mimir-2 Kube state metrices available
<img width="150" alt="image" src="https://github.com/grafana/mimir/assets/111333290/9ebda885-7461-4cf7-9340-8f2473d053f0">

Regarding the components Mimir-1 we are using only the required components and optional are not used

**Reply by dimitarvdimitrov** [reply]
there are three possibilities that I can think of
* Mimir is rejecting the kube state metrics from the collectors due to some problem with them (duplicate labels, too old, too new, etc...)
	* If mimir is rejecting the requests, then you might be able to find rejection errors in the distributor logs or in the otel collector logs (I'm not sure about the latter, I assume there are some kind logs). The distributor logs will contain strings such as `err-mimir-max-label-names-per-series`. You can find the full catalog [here](https://grafana.com/docs/mimir/latest/operators-guide/mimir-runbooks/#errors-catalog).
* The otel collector isn't sending the kube-state-metrics to Mimir
	* I'm not that familiar with the otel collector, but is there a way to verify what metrics are being scraped and remote-written by the collector (ideally as two separate sets of metrics). That will narrow down the problem to either the collector or Mimir.
* The otel collector is sending the KSM metrics to Mimir, but they end up under a separate tenant
	* you can review the collector configuration for the two clusters and verify they are using the same `X-Scope-OrgID` HTTP header for KSM

**Reply by akalita-conga** [reply]
Mimir distributor current logs --
```
ts=2023-07-12T12:54:53.263634953Z caller=main.go:216 level=info msg="Starting application" version="(version=2.9.0, branch=HEAD, revision=761114d)"
ts=2023-07-12T12:54:53.265190743Z caller=server.go:336 level=info http=[::]:8080 grpc=[::]:9095 msg="server listening on addresses"
ts=2023-07-12T12:54:53.270915725Z caller=memberlist_client.go:437 level=info msg="Using memberlist cluster label and node name" cluster_label= node=mimir-infra-distributor-6fbf694984-7rvzv-1c63ee6e
ts=2023-07-12T12:54:53.273305275Z caller=module_service.go:82 level=info msg=initialising module=activity-tracker
ts=2023-07-12T12:54:53.273357451Z caller=module_service.go:82 level=info msg=initialising module=sanity-check
ts=2023-07-12T12:54:53.273375781Z caller=module_service.go:82 level=info msg=initialising module=active-groups-cleanup-service
ts=2023-07-12T12:54:53.273394398Z caller=sanity_check.go:32 level=info msg="Checking directories read/write access"
ts=2023-07-12T12:54:53.273414632Z caller=sanity_check.go:37 level=info msg="Directories read/write access successfully checked"
ts=2023-07-12T12:54:53.273427218Z caller=sanity_check.go:39 level=info msg="Checking object storage config"
ts=2023-07-12T12:54:53.273457889Z caller=sanity_check.go:44 level=info msg="Object storage config successfully checked"
ts=2023-07-12T12:54:53.273473758Z caller=module_service.go:82 level=info msg=initialising module=server
ts=2023-07-12T12:54:53.273502951Z caller=module_service.go:82 level=info msg=initialising module=memberlist-kv
ts=2023-07-12T12:54:53.273516834Z caller=module_service.go:82 level=info msg=initialising module=runtime-config
ts=2023-07-12T12:54:53.273715422Z caller=module_service.go:82 level=info msg=initialising module=ring
ts=2023-07-12T12:54:53.277938504Z caller=memberlist_client.go:543 level=info msg="memberlist fast-join starting" nodes_found=10 to_join=8
ts=2023-07-12T12:54:53.289008975Z caller=memberlist_client.go:563 level=info msg="memberlist fast-join finished" joined_nodes=8 elapsed_time=16.658862ms
ts=2023-07-12T12:54:53.289076002Z caller=memberlist_client.go:576 level=info msg="joining memberlist cluster" join_members=dns+mimir-infra-gossip-ring.monitoring.svc.cluster.local:7946
ts=2023-07-12T12:54:53.289492935Z caller=module_service.go:82 level=info msg=initialising module=distributor-service
ts=2023-07-12T12:54:53.289638747Z caller=basic_lifecycler.go:276 level=info msg="instance not found in the ring" instance=mimir-infra-distributor-6fbf694984-7rvzv ring=distributor
ts=2023-07-12T12:54:53.289809692Z caller=distributor.go:464 level=info msg="waiting until distributor is ACTIVE in the ring"
ts=2023-07-12T12:54:53.289905897Z caller=mimir.go:825 level=info msg="Application started"
ts=2023-07-12T12:54:56.439701569Z caller=memberlist_client.go:595 level=info msg="joining memberlist cluster succeeded" reached_nodes=9 elapsed_time=3.150625599s
```

**Reply by dimitarvdimitrov** [reply]
there's nothing out of the ordinary in these logs. This rules out 

> * Mimir is rejecting the kube state metrics from the collectors due to some problem with them (duplicate labels, too old, too new, etc...)
>   * If mimir is rejecting the requests, then you might be able to find rejection errors in the distributor logs or in the otel collector logs (I'm not sure about the latter, I assume there are some kind logs). The distributor logs will contain strings such as `err-mimir-max-label-names-per-series`. You can find the full catalog [here](https://grafana.com/docs/mimir/latest/operators-guide/mimir-runbooks/#errors-catalog).

**Reply by akalita-conga** [reply]
@dimitarvdimitrov :  There is no related error found as said.

**Reply by dimitarvdimitrov** [reply]
Have you had a chance to look into the other two possible causes?

> * The otel collector isn't sending the kube-state-metrics to Mimir
>   
>   * I'm not that familiar with the otel collector, but is there a way to verify what metrics are being scraped and remote-written by the collector (ideally as two separate sets of metrics). That will narrow down the problem to either the collector or Mimir.
> * The otel collector is sending the KSM metrics to Mimir, but they end up under a separate tenant
>   
>   * you can review the collector configuration for the two clusters and verify they are using the same `X-Scope-OrgID` HTTP header for KSM

**Reply by akalita-conga** [reply]
@dimitarvdimitrov 
 There is no specific error in the logs .One thing I observe is one Mimir is getting metrices from prometheus /push as well as Otel/ push and another one is getting only otel push .
  I wonder how prometheus is pushing metrices to mimir-2 as we don't have prometheus deployed in the cluster.We are receiving all the metrices from the app cluster. 

Sample Nginx logs 
The Mimir-2 is getting metrics from a prometheus endpoint as well as Opentelemetry

 

 10.168.129.134 - - [17/Jul/2023:08:24:26 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "Prometheus/2.35.0" "-"
10.168.129.134 - - [17/Jul/2023:08:24:26 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "Prometheus/2.35.0" "-"
10.168.129.134 - - [17/Jul/2023:08:24:26 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "Prometheus/2.35.0" "-"
10.168.129.134 - - [17/Jul/2023:08:24:27 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "Prometheus/2.35.0" "-"
10.168.129.134 - - [17/Jul/2023:08:24:27 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "Prometheus/2.35.0" "-"
10.168.129.134 - - [17/Jul/2023:08:24:27 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "Prometheus/2.35.0" 

 

Mimir-1 is getting metrics from only from opentelemetry

 

 10.168.189.12 - - [17/Jul/2023:08:23:26 +0000]  200 "GET / HTTP/1.1" 2 "-" "kube-probe/1.23+" "-"
10.168.174.20 - - [17/Jul/2023:08:23:33 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "opentelemetry-collector-contrib/0.73.0" "-"
10.168.162.244 - - [17/Jul/2023:08:23:35 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "opentelemetry-collector-contrib/0.73.0" "-"
10.168.189.12 - - [17/Jul/2023:08:23:36 +0000]  200 "GET / HTTP/1.1" 2 "-" "kube-probe/1.23+" "-"
10.168.174.20 - - [17/Jul/2023:08:23:43 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "opentelemetry-collector-contrib/0.73.0" "-"
10.168.162.244 - - [17/Jul/2023:08:23:45 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "opentelemetry-collector-contrib/0.73.0" "-"
10.168.189.12 - - [17/Jul/2023:08:23:46 +0000]  200 "GET / HTTP/1.1" 2 "-" "kube-probe/1.23+" "-"
10.168.174.20 - - [17/Jul/2023:08:23:53 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "opentelemetry-collector-contrib/0.73.0" "-"
10.168.162.244 - - [17/Jul/2023:08:23:55 +0000]  200 "POST /api/v1/push HTTP/1.1" 0 "-" "opentelemetry-collector-contrib/0.73.0" "-"
